scertecd/scertecd.go

```diff
@@ -0,0 +1,862 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// The scertecd package provides the code that fetches new TLS certs
+// from LetsEncrypt as needed and puts them in setec before they
+// expire. The code can run either in the foreground once, or most
+// commonly as an HTTP server daemon.
+//
+// It populates the following setec keys:
+//
+//   - {prod,dev}/scertec/acme-key: the private key for the ACME account, as a PEM-encoded ECDSA key
+//   - {prod,dev}/scertec/https-cert/{domain-name}/rsa: PEM of private key, domain cert, LetsEncrypt cert
+//   - {prod,dev}/scertec/https-cert/{domain-name}/ecdsa: PEM of private key, domain cert, LetsEncrypt cert
+package scertecd
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"html/template"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/tailscale/setec/client/setec"
+	"github.com/tailscale/setec/types/api"
+	"golang.org/x/crypto/acme"
+)
+
+// Server is the scertec updater server.
+//
+// Despite the name "server", it can also be used in a single-shot
+// foreground mode via its UpdateAll method.
+//
+// All exported fields must be initialized before calling an exported
+// method on the Server: either UpdateAll or Start.
+type Server struct {
+	SetecClient setec.Client                     // required client for setec
+	Domains     []string                         // domains to maintain certs for
+	Now         func() time.Time                 // if nil, time.Now is used
+	ACMEContact string                           // optional email address for ACME registration
+	Prefix      string                           // setec secret prefix
+	Logf        func(format string, args ...any) // alteranate log function; if nil, log.Printf is used
+
+	dtsOnce sync.Once
+	dts     []domainAndType
+
+	mu       sync.Mutex
+	acLazy   *acme.Client // nil until needed via getACMEClient
+	last     map[domainAndType]*certUpdateCheck
+	secCache map[string]*api.SecretValue
+}
+
+func (s *Server) initDomainAndTypes() {
+	s.dtsOnce.Do(func() {
+		for _, d := range s.Domains {
+			for _, typ := range []CertType{RSACert, ECDSACert} {
+				s.dts = append(s.dts, domainAndType{d, typ})
+			}
+		}
+	})
+}
+
+func (s *Server) logf(format string, args ...any) {
+	if s.Logf != nil {
+		s.Logf(format, args...)
+	} else {
+		log.Printf(format, args...)
+	}
+}
```

Comment:
maybe initialize `s.Logf` when nil in `initDomainAndTypes` or a similar method?
it's too easy to call `s.Logf(...)` instead of `s.logf(...)` and cause a nil pointer panic

Reply:
done